src/main/java/fi/helsinki/cs/tmc/core/domain/Course.java

```diff
@@ -67,6 +71,39 @@ public void setExercises(List<Exercise> exercises) {
         this.exercises = exercises;
     }
 
+    public List<Theme> getThemes() {
+        return themes;
+    }
+
+    public void generateThemes() {
+        themes = new ArrayList<>();
+        Map<String, Theme> themeMap = new HashMap<>();
+        for (Exercise ex : exercises) {
+            addExerciseToTheme(themeMap, ex);
+        }
+    }
+
+    private void addExerciseToTheme(Map<String, Theme> themeMap, Exercise ex) {
+        String themeName = ex.getName().split("-")[0];
+        Theme theme = themeMap.get(themeName);
+        if (theme == null) {
+            theme = new Theme(themeName);
+            themeMap.put(themeName, theme);
+            themes.add(theme);
+        }
+        theme.addExercise(ex);
+    }
+
+    public List<Exercise> getExercisesByTheme(Theme theme) {
+        ArrayList<Exercise> exercisesByTheme = new ArrayList<>();
+        for (Exercise exercise : theme.getExercises()) {
+            if (this.exercises.contains(exercise)) {
+                exercisesByTheme.add(exercise);
+            }
+        }
+        return exercisesByTheme;
+    }
```

Comment:
**This comment is not a request for change, I'm just trying to get a better understanding of the big picture.**

This method seems to just create the set intersection between the exercises in `theme` and the exercises in `this` course.

I don't fully understand why this is useful: To get all exercises that relate to a certain `Theme` within a certain `Course`, you would already need to have the `Theme` contain all the exercises.

This architecture might be completely valid in the larger picture I'm not seeing, but to me it seems that the signature `public List<Exercise> getExercisesByThemeName(String string)` or something similar would be more helpful.